File: src/org/usfirst/frc/team4571/robot/subsystems/Elevator.java
package org.usfirst.frc.team4571.robot.subsystems;

import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.PIDController;
import edu.wpi.first.wpilibj.command.Subsystem;
import org.usfirst.frc.team4571.robot.Robot;
import org.usfirst.frc.team4571.robot.RobotMap;
import org.usfirst.frc.team4571.robot.subsystems.pid.ElevatorEncoder;
import org.usfirst.frc.team4571.robot.subsystems.pid.ElevatorOutput;

public class Elevator extends Subsystem {

    private static final double
            MAX_HEIGHT_ENCODER_TICK = 29000,
            MAX_HEIGHT_IN_INCHES = 89;
    private static final double
            p = 0.5,
            i = 0,
            d = 0;
    private final PIDController elevatorController;
    private final ElevatorEncoder elevatorEncoder;
    private WPI_TalonSRX elevatorMotor;
    private DigitalInput lowerLimit;

    public Elevator() {
        this.elevatorMotor = new WPI_TalonSRX(RobotMap.ELEVATOR_MOTOR);
        this.elevatorMotor.setExpiration(Robot.DEFAULT_PERIOD);
        this.elevatorMotor.setSafetyEnabled(false);
        this.elevatorMotor.setNeutralMode(NeutralMode.Brake);

        this.elevatorEncoder = new ElevatorEncoder(elevatorMotor);
        this.elevatorController = new PIDController(
                p, i, d, elevatorEncoder, new ElevatorOutput(elevatorMotor));

        this.lowerLimit = new DigitalInput(RobotMap.LIMIT_SWITCH);
    }

    public void initDefaultCommand() {}

    /**
     * converts the given height in inches to encoder ticks
     *
     * @param setpointInInches setpoint in inches
     */
    private int convertSetpoint(double setpointInInches) {
        return (int) ((setpointInInches / MAX_HEIGHT_IN_INCHES) *
                      MAX_HEIGHT_ENCODER_TICK);
    }

    public double getTick() {
        return elevatorEncoder.pidGet();
    }

    public PIDController getElevatorController() {
        return elevatorController;
    }

    public void setHeight(double heightInInches) {
        elevatorController.setInputRange(0, MAX_HEIGHT_ENCODER_TICK);
        elevatorController.setOutputRange(-0.5, 0.8);
        elevatorController.setSetpoint(
                getTick() + convertSetpoint(heightInInches));
        elevatorController.setAbsoluteTolerance(
                heightInInches * 0.1); // 10% tolerance
        elevatorController.enable();
    }

    public void stopElevator() {
        elevatorMotor.set(0.0);
    }

    public void resetEncoder() {
        elevatorEncoder.reset();
    }

    public void disablePID() {
        this.elevatorController.disable();
    }

    public double getElevatorSpeed() {
        return elevatorMotor.get();
    }

    public void setElevatorMotor(double power) {
        elevatorMotor.set(power);
    }

    public boolean isOnTarget() {
        return elevatorController.onTarget();
    }

    public boolean isLimitSwitchPressed() {
        return lowerLimit.get();
    }
}

File: src/org/usfirst/frc/team4571/robot/subsystems/Elevator.kt
package org.usfirst.frc.team4571.robot.subsystems

import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX
import edu.wpi.first.wpilibj.DigitalInput
import edu.wpi.first.wpilibj.PIDController
import edu.wpi.first.wpilibj.PIDOutput
import edu.wpi.first.wpilibj.command.Subsystem
import org.usfirst.frc.team4571.robot.RobotMap
import org.usfirst.frc.team4571.robot.hardware.CanTalon
import org.usfirst.frc.team4571.robot.hardware.ElevatorEncoder

object Elevator : Subsystem() {
    private const val MAX_HEIGHT_ENCODER_TICK = 29000.0
    private const val MAX_HEIGHT_IN_INCHES = 89.0

    private val elevatorMotor: WPI_TalonSRX = CanTalon(RobotMap.Elevator.ELEVATOR_MOTOR)
    private val lowerLimit: DigitalInput = DigitalInput(RobotMap.Elevator.LIMIT_SWITCH)
    private val elevatorEncoder: ElevatorEncoder = ElevatorEncoder(elevatorMotor)
    val elevatorController: PIDController =
        PIDController(Encoder.P, Encoder.I, Encoder.D, elevatorEncoder, ElevatorOutput)

    val tick: Double
        get() = elevatorEncoder.pidGet()

    val elevatorSpeed: Double
        get() = elevatorMotor.get()

    val isOnTarget: Boolean
        get() = this.elevatorController.onTarget()

    val isLimitSwitchPressed: Boolean
        get() = lowerLimit.get()


    private object Encoder {
        const val P = 0.5
        const val I = 0.0
        const val D = 0.0
    }

    init {
        elevatorController.setInputRange(0.0, MAX_HEIGHT_ENCODER_TICK)
        elevatorController.setOutputRange(-0.5, 0.8)
    }

    private object ElevatorOutput : PIDOutput {
        override fun pidWrite(output: Double) = setElevatorMotor(output)
    }

    public override fun initDefaultCommand() {}

    fun setElevatorMotor(power: Double) = elevatorMotor.set(power)

    private fun convertSetpoint(setpointInInches: Double): Int {
        return (setpointInInches / MAX_HEIGHT_IN_INCHES *
                MAX_HEIGHT_ENCODER_TICK).toInt()
    }

    fun goToHeight(height: Double) {
        elevatorEncoder.reset()
        elevatorController.setpoint = convertSetpoint(height).toDouble()
        elevatorController.setAbsoluteTolerance(height * 0.1) // 10% tolerance
        elevatorController.enable()
    }

    fun stopElevator() = elevatorMotor.stopMotor()

    fun resetEncoder() = elevatorEncoder.reset()

    fun disablePID() = elevatorController.disable()
}